File: app/frontend/src/lib/lakebaseEstimate.ts
/**
 * Lakebase CU / cost estimation from CPU utilization samples.
 * Data-only — no chart formatting or downsampling; callers own UI/chart prep.
 */
import type { MetricDataPoint, MetricResponse } from "../types";

/** Matches single-server Lakebase estimate dialog defaults. */
export const LAKEBASE_ESTIMATE_DEFAULT_SAFETY_MARGIN_PCT = 15;

/** "Scale to zero" when idle — same checkbox as individual server estimate. */
export const LAKEBASE_ESTIMATE_DEFAULT_SCALE_TO_ZERO = true;

/** Lakebase compute units per physical core used in the estimate formula. */
export const LAKEBASE_CU_PER_CORE = 4;

/**
 * If any interval’s required CU (from peak CPU% × vCores, with safety margin)
 * reaches this level, scale-to-zero is not applied and monthly CU uses the
 * **peak** interval CU × periods/month (provisioned-style sizing).
 */
export const LAKEBASE_CU_HIGH_USAGE_THRESHOLD = 32;

/** Cores below this (when scale-to-zero is on) map to 0 CU for that period. */
export const LAKEBASE_SCALE_TO_ZERO_THRESHOLD_CORES = 0.3;

/** Hours per month used to project sampled periods to monthly CU. */
export const LAKEBASE_HOURS_PER_MONTH = 730;

/** USD per Lakebase CU-month (placeholder — align with your pricing sheet). */
export const LAKEBASE_CU_USD_PER_UNIT = 0.111;

/** Compute discount when scale-to-zero does not reduce any interval (100% uptime). */
export const LAKEBASE_100_PERCENT_UPTIME_DISCOUNT_PCT = 25;

/** Storage $/GB-month when SKU looks like AWS RDS (db.*). */
export const LAKEBASE_STORAGE_USD_PER_GB_AWS = 0.345;

/** Storage $/GB-month default (e.g. Azure / other). */
export const LAKEBASE_STORAGE_USD_PER_GB_DEFAULT = 0.39;

export function isAwsSku(skuName: string | null | undefined): boolean {
  return skuName != null && skuName.startsWith("db.");
}

export function lakebaseStorageUsdPerGb(
  skuName: string | null | undefined
): number {
  return isAwsSku(skuName)
    ? LAKEBASE_STORAGE_USD_PER_GB_AWS
    : LAKEBASE_STORAGE_USD_PER_GB_DEFAULT;
}

export interface LakebaseEstimateOptions {
  safetyMarginPct: number;
  scaleToZero: boolean;
}

/** One time-aligned row after CPU% → cores and CU math. */
export interface LakebaseEstimatePoint {
  timestamp: string;
  coresUsedAvg: number | null;
  coresUsedMax: number | null;
  lakebaseCU: number | null;
}

/** Rolled-up numbers for summaries and copy. */
export interface LakebaseEstimateMetrics {
  peakCores: number;
  avgCores: number;
  safetyLineCores: number;
  monthlyCU: number;
  scaleToZeroPeriods: number;
  totalPeriods: number;
  intervalHours: number;
  periodsPerMonth: number;
  avgCUPerPeriod: number;
  /** True when any interval hit {@link LAKEBASE_CU_HIGH_USAGE_THRESHOLD}+ CUs; S2Z ignored; monthly from peak interval CU. */
  usedPeakCuConstantSizing: boolean;
  /** Largest per-interval Lakebase CU in the final model (after rules). */
  peakPeriodLakebaseCU: number;
  /** True when scale-to-zero is off or no intervals qualify for 0 CU. */
  qualifiesFor100PercentUptimeDiscount: boolean;
}

export interface LakebaseEstimateResult {
  points: LakebaseEstimatePoint[];
  metrics: LakebaseEstimateMetrics;
}

/**
 * Build per-period points and aggregate metrics from CPU % samples and vCore count.
 */
export function computeLakebaseEstimate(
  cpuData: MetricDataPoint[],
  vcores: number,
  options: LakebaseEstimateOptions
): LakebaseEstimateResult {
  const { safetyMarginPct, scaleToZero } = options;
  const margin = safetyMarginPct / 100;
  const idleThresholdCores = LAKEBASE_SCALE_TO_ZERO_THRESHOLD_CORES;

  const anyPeriodRequiresHighCu = cpuData.some((d) => {
    if (d.maximum == null) return false;
    const maxUsed = (d.maximum / 100) * vcores;
    const cu = Math.ceil(maxUsed * (1 + margin) * LAKEBASE_CU_PER_CORE);
    return cu >= LAKEBASE_CU_HIGH_USAGE_THRESHOLD;
  });

  const effectiveScaleToZero = anyPeriodRequiresHighCu ? false : scaleToZero;

  let totalCores = 0;
  let count = 0;
  let peak = 0;

  const points: LakebaseEstimatePoint[] = cpuData.map((d) => {
    const avgUsed =
      d.average != null ? (d.average / 100) * vcores : null;
    const maxUsed =
      d.maximum != null ? (d.maximum / 100) * vcores : null;

    if (avgUsed != null) {
      totalCores += avgUsed;
      count++;
    }
    if (maxUsed != null && maxUsed > peak) {
      peak = maxUsed;
    }

    const isIdle =
      effectiveScaleToZero && maxUsed != null && maxUsed < idleThresholdCores;
    const cuRequired =
      maxUsed != null
        ? isIdle
          ? 0
          : Math.ceil(maxUsed * (1 + margin) * LAKEBASE_CU_PER_CORE)
        : null;

    return {
      timestamp: d.timestamp,
      coresUsedAvg:
        avgUsed != null ? Math.round(avgUsed * 100) / 100 : null,
      coresUsedMax:
        maxUsed != null ? Math.round(maxUsed * 100) / 100 : null,
      lakebaseCU: cuRequired,
    };
  });

  const avg = count > 0 ? totalCores / count : 0;
  const safetyCores = peak * (1 + margin);

  let totalCU = 0;
  let cuCount = 0;
  let s2zPeriods = 0;
  for (const p of points) {
    if (p.lakebaseCU != null) {
      totalCU += p.lakebaseCU;
      cuCount++;
      if (p.lakebaseCU === 0 && effectiveScaleToZero) {
        s2zPeriods++;
      }
    }
  }

  const finiteCUs = points
    .map((p) => p.lakebaseCU)
    .filter((c): c is number => c != null);
  const peakPeriodLakebaseCU =
    finiteCUs.length > 0 ? Math.max(...finiteCUs) : 0;

  let intervalHours = 1;
  if (cpuData.length >= 2) {
    const firstTs = new Date(cpuData[0].timestamp).getTime();
    const lastTs = new Date(cpuData[cpuData.length - 1].timestamp).getTime();
    intervalHours =
      (lastTs - firstTs) / (cpuData.length - 1) / 3600000;
  }

  const periodsPerMonth = LAKEBASE_HOURS_PER_MONTH / intervalHours;

  let monthlyCU: number;
  let avgCUPerPeriod: number;
  if (anyPeriodRequiresHighCu) {
    avgCUPerPeriod = peakPeriodLakebaseCU;
    monthlyCU = Math.round(peakPeriodLakebaseCU * periodsPerMonth);
  } else {
    avgCUPerPeriod = cuCount > 0 ? totalCU / cuCount : 0;
    monthlyCU = Math.round(avgCUPerPeriod * periodsPerMonth);
  }

  const metrics: LakebaseEstimateMetrics = {
    peakCores: Math.round(peak * 100) / 100,
    avgCores: Math.round(avg * 100) / 100,
    safetyLineCores: Math.round(safetyCores * 100) / 100,
    monthlyCU,
    scaleToZeroPeriods: s2zPeriods,
    totalPeriods: cuCount,
    intervalHours,
    periodsPerMonth,
    avgCUPerPeriod,
    usedPeakCuConstantSizing: anyPeriodRequiresHighCu,
    peakPeriodLakebaseCU,
    qualifiesFor100PercentUptimeDiscount:
      qualifiesFor100PercentUptimeDiscount(scaleToZero, s2zPeriods),
  };

  return { points, metrics };
}

/** When scale-to-zero is unchecked or no intervals map to 0 CU. */
export function qualifiesFor100PercentUptimeDiscount(
  scaleToZeroRequested: boolean,
  scaleToZeroPeriods: number
): boolean {
  return !scaleToZeroRequested || scaleToZeroPeriods === 0;
}

export function apply100PercentUptimeDiscountUsd(amountUsd: number): number {
  return amountUsd * (1 - LAKEBASE_100_PERCENT_UPTIME_DISCOUNT_PCT / 100);
}

export function lakebaseMonthlyCuCostUsdAfterUptimeDiscount(
  monthlyCU: number,
  qualifiesForDiscount: boolean
): number {
  const base = lakebaseMonthlyCuCostUsd(monthlyCU);
  return qualifiesForDiscount
    ? apply100PercentUptimeDiscountUsd(base)
    : base;
}

export function lakebaseTotalMonthlyCostUsdAfterUptimeDiscount(
  monthlyCU: number,
  storageGb: number | null | undefined,
  skuName: string | null | undefined,
  qualifiesForDiscount: boolean
): number {
  return (
    lakebaseMonthlyCuCostUsdAfterUptimeDiscount(monthlyCU, qualifiesForDiscount) +
    lakebaseStorageMonthlyCostUsd(storageGb, skuName)
  );
}

/**
 * Same inputs as the individual analysis view: full metrics list, vCores, and
 * {@link LakebaseEstimateOptions} (safety margin + scaleToZero / "autoscale when idle").
 */
export function tryComputeLakebaseEstimateFromMetrics(
  allMetrics: MetricResponse[],
  vcores: number | null | undefined,
  options: LakebaseEstimateOptions
):
  | { ok: true; result: LakebaseEstimateResult; cpuMetric: MetricResponse }
  | { ok: false; error: string } {
  if (vcores == null || vcores <= 0) {
    return { ok: false, error: "vCores missing or invalid" };
  }
  const cpu = allMetrics.find((m) => m.metric_name === "cpu_percent");
  if (!cpu || cpu.data_points === 0) {
    return { ok: false, error: "No CPU percent metric data" };
  }
  const result = computeLakebaseEstimate(cpu.data, vcores, options);
  return { ok: true, result, cpuMetric: cpu };
}

export function lakebaseMonthlyCuCostUsd(monthlyCU: number): number {
  return monthlyCU * LAKEBASE_CU_USD_PER_UNIT;
}

export function lakebaseStorageMonthlyCostUsd(
  storageGb: number | null | undefined,
  skuName: string | null | undefined
): number {
  if (storageGb == null) return 0;
  return storageGb * lakebaseStorageUsdPerGb(skuName);
}

/** For branched Lakebase databases, sizing uses this fraction of reported storage. */
export const LAKEBASE_BRANCHED_STORAGE_FRACTION = 0.1;

/**
 * Storage GB used for Lakebase storage $/mo (branched DBs count only a fraction of reported size).
 */
export function effectiveStorageGbForLakebaseSizing(
  storageGb: number | null | undefined,
  isBranchedDatabase: boolean
): number | null {
  if (storageGb == null) return null;
  if (!isBranchedDatabase) return storageGb;
  return Math.max(0, storageGb * LAKEBASE_BRANCHED_STORAGE_FRACTION);
}

export function lakebaseTotalMonthlyCostUsd(
  monthlyCU: number,
  storageGb: number | null | undefined,
  skuName: string | null | undefined
): number {
  return (
    lakebaseMonthlyCuCostUsd(monthlyCU) +
    lakebaseStorageMonthlyCostUsd(storageGb, skuName)
  );
}

File: app/mcp_server/lakebase_estimate.py
"""
Lakebase CU / cost estimation (Python port of app/frontend/src/lib/lakebaseEstimate.ts).
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import datetime
from typing import Any

LAKEBASE_CU_PER_CORE = 4
LAKEBASE_CU_HIGH_USAGE_THRESHOLD = 32
LAKEBASE_SCALE_TO_ZERO_THRESHOLD_CORES = 0.3
LAKEBASE_HOURS_PER_MONTH = 730
LAKEBASE_CU_USD_PER_UNIT = 0.111
LAKEBASE_100_PERCENT_UPTIME_DISCOUNT_PCT = 25
LAKEBASE_STORAGE_USD_PER_GB_AWS = 0.345
LAKEBASE_STORAGE_USD_PER_GB_DEFAULT = 0.39
LAKEBASE_BRANCHED_STORAGE_FRACTION = 0.1


def is_aws_sku(sku_name: str | None) -> bool:
    return sku_name is not None and sku_name.startswith("db.")


def lakebase_storage_usd_per_gb(sku_name: str | None) -> float:
    return (
        LAKEBASE_STORAGE_USD_PER_GB_AWS
        if is_aws_sku(sku_name)
        else LAKEBASE_STORAGE_USD_PER_GB_DEFAULT
    )


@dataclass
class MetricDataPoint:
    timestamp: str
    average: float | None
    maximum: float | None
    minimum: float | None


@dataclass
class LakebaseEstimatePoint:
    timestamp: str
    cores_used_avg: float | None
    cores_used_max: float | None
    lakebase_cu: int | None


@dataclass
class LakebaseEstimateMetrics:
    peak_cores: float
    avg_cores: float
    safety_line_cores: float
    monthly_cu: int
    scale_to_zero_periods: int
    total_periods: int
    interval_hours: float
    periods_per_month: float
    avg_cu_per_period: float
    used_peak_cu_constant_sizing: bool
    peak_period_lakebase_cu: int
    qualifies_for_100_percent_uptime_discount: bool


@dataclass
class LakebaseEstimateResult:
    points: list[LakebaseEstimatePoint]
    metrics: LakebaseEstimateMetrics


def compute_lakebase_estimate(
    cpu_data: list[MetricDataPoint],
    vcores: int,
    safety_margin_pct: float,
    scale_to_zero: bool,
) -> LakebaseEstimateResult:
    margin = safety_margin_pct / 100.0
    idle_threshold = LAKEBASE_SCALE_TO_ZERO_THRESHOLD_CORES

    any_period_requires_high_cu = False
    for d in cpu_data:
        if d.maximum is None:
            continue
        max_used = (d.maximum / 100.0) * vcores
        cu = int(math.ceil(max_used * (1 + margin) * LAKEBASE_CU_PER_CORE))
        if cu >= LAKEBASE_CU_HIGH_USAGE_THRESHOLD:
            any_period_requires_high_cu = True
            break

    effective_scale_to_zero = False if any_period_requires_high_cu else scale_to_zero

    total_cores = 0.0
    count = 0
    peak = 0.0
    points: list[LakebaseEstimatePoint] = []

    for d in cpu_data:
        avg_used = (d.average / 100.0) * vcores if d.average is not None else None
        max_used = (d.maximum / 100.0) * vcores if d.maximum is not None else None

        if avg_used is not None:
            total_cores += avg_used
            count += 1
        if max_used is not None and max_used > peak:
            peak = max_used

        is_idle = (
            effective_scale_to_zero
            and max_used is not None
            and max_used < idle_threshold
        )
        if max_used is not None:
            cu_required = (
                0
                if is_idle
                else int(math.ceil(max_used * (1 + margin) * LAKEBASE_CU_PER_CORE))
            )
        else:
            cu_required = None

        points.append(
            LakebaseEstimatePoint(
                timestamp=d.timestamp,
                cores_used_avg=round(avg_used, 2) if avg_used is not None else None,
                cores_used_max=round(max_used, 2) if max_used is not None else None,
                lakebase_cu=cu_required,
            )
        )

    avg = total_cores / count if count > 0 else 0.0
    safety_cores = peak * (1 + margin)

    total_cu = 0
    cu_count = 0
    s2z_periods = 0
    for p in points:
        if p.lakebase_cu is not None:
            total_cu += p.lakebase_cu
            cu_count += 1
            if p.lakebase_cu == 0 and effective_scale_to_zero:
                s2z_periods += 1

    finite_cus = [p.lakebase_cu for p in points if p.lakebase_cu is not None]
    peak_period_lakebase_cu = max(finite_cus) if finite_cus else 0

    interval_hours = 1.0
    if len(cpu_data) >= 2:

        def _ts_to_epoch(ts: str) -> float:
            t = ts.strip()
            if t.endswith("Z"):
                t = t[:-1] + "+00:00"
            try:
                return datetime.fromisoformat(t).timestamp()
            except ValueError:
                return datetime.fromisoformat(t + "+00:00").timestamp()

        first_ts = _ts_to_epoch(cpu_data[0].timestamp)
        last_ts = _ts_to_epoch(cpu_data[-1].timestamp)
        interval_hours = (last_ts - first_ts) / (len(cpu_data) - 1) / 3600.0

    periods_per_month = LAKEBASE_HOURS_PER_MONTH / interval_hours

    if any_period_requires_high_cu:
        avg_cu_per_period = float(peak_period_lakebase_cu)
        monthly_cu = int(round(peak_period_lakebase_cu * periods_per_month))
    else:
        avg_cu_per_period = total_cu / cu_count if cu_count > 0 else 0.0
        monthly_cu = int(round(avg_cu_per_period * periods_per_month))

    metrics = LakebaseEstimateMetrics(
        peak_cores=round(peak, 2),
        avg_cores=round(avg, 2),
        safety_line_cores=round(safety_cores, 2),
        monthly_cu=monthly_cu,
        scale_to_zero_periods=s2z_periods,
        total_periods=cu_count,
        interval_hours=interval_hours,
        periods_per_month=periods_per_month,
        avg_cu_per_period=avg_cu_per_period,
        used_peak_cu_constant_sizing=any_period_requires_high_cu,
        peak_period_lakebase_cu=peak_period_lakebase_cu,
        qualifies_for_100_percent_uptime_discount=qualifies_for_100_percent_uptime_discount(
            scale_to_zero, s2z_periods
        ),
    )

    return LakebaseEstimateResult(points=points, metrics=metrics)


def qualifies_for_100_percent_uptime_discount(
    scale_to_zero_requested: bool, scale_to_zero_periods: int
) -> bool:
    return not scale_to_zero_requested or scale_to_zero_periods == 0


def apply_100_percent_uptime_discount_usd(amount_usd: float) -> float:
    return amount_usd * (1 - LAKEBASE_100_PERCENT_UPTIME_DISCOUNT_PCT / 100.0)


def monthly_cu_cost_usd(monthly_cu: int) -> float:
    return monthly_cu * LAKEBASE_CU_USD_PER_UNIT


def monthly_cu_cost_usd_after_uptime_discount(
    monthly_cu: int, qualifies_for_discount: bool
) -> float:
    base = monthly_cu_cost_usd(monthly_cu)
    return (
        apply_100_percent_uptime_discount_usd(base)
        if qualifies_for_discount
        else base
    )


def storage_monthly_cost_usd(storage_gb: int | None, sku_name: str | None) -> float:
    if storage_gb is None:
        return 0.0
    return storage_gb * lakebase_storage_usd_per_gb(sku_name)


def effective_storage_gb_for_lakebase_sizing(
    storage_gb: int | None, is_branched_database: bool
) -> int | None:
    if storage_gb is None:
        return None
    if not is_branched_database:
        return storage_gb
    return max(0, int(round(storage_gb * LAKEBASE_BRANCHED_STORAGE_FRACTION)))


def total_monthly_cost_usd(
    monthly_cu: int, storage_gb: int | None, sku_name: str | None
) -> float:
    return monthly_cu_cost_usd(monthly_cu) + storage_monthly_cost_usd(
        storage_gb, sku_name
    )


def total_monthly_cost_usd_after_uptime_discount(
    monthly_cu: int,
    storage_gb: int | None,
    sku_name: str | None,
    qualifies_for_discount: bool,
) -> float:
    return monthly_cu_cost_usd_after_uptime_discount(
        monthly_cu, qualifies_for_discount
    ) + storage_monthly_cost_usd(storage_gb, sku_name)


def metrics_rows_to_cpu_points(rows: list[dict[str, Any]]) -> list[MetricDataPoint]:
    out: list[MetricDataPoint] = []
    for r in rows:
        out.append(
            MetricDataPoint(
                timestamp=str(r["timestamp"]),
                average=r.get("average"),
                maximum=r.get("maximum"),
                minimum=r.get("minimum"),
            )
        )
    return out
